kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+             * information.
+             *
+             * Requires that `filesIter.hasNext` is true.
+             */
+            private CloseableIterator<Tuple2<FileDataReadResult, Boolean>> getNextActionsIter() {
+                final FileStatus nextFile = filesIter.next();
+                final Path nextFilePath = new Path(nextFile.getPath());
+
+                try {
+                    if (nextFilePath.getName().endsWith(".json")) {
+                        final JsonHandler jsonHandler = tableClient.getJsonHandler();
+
+                        // Convert the `nextFile` FileStatus into an internal ScanFile Row and then
+                        // allow the connector to contextualize it (i.e. perform any splitting)
+                        final CloseableIterator<FileReadContext> fileReadContextIter =
+                            jsonHandler.contextualizeFileReads(
+                                Utils.singletonCloseableIterator(
+                                    InternalUtils.getScanFileRow(nextFile)),
+                                Literal.TRUE
+                            );
+
+                        // Read that file
+                        final CloseableIterator<FileDataReadResult> fileReadDataIter =
+                            tableClient.getJsonHandler().readJsonFiles(
+                                fileReadContextIter,
+                                readSchema);
+
+                        return combine(fileReadDataIter, false /* isFromCheckpoint */);
+                    } else if (nextFilePath.getName().endsWith(".parquet")) {
+                        final ParquetHandler parquetHandler = tableClient.getParquetHandler();
+
+                        // Convert the `nextFile` FileStatus into an internal ScanFile Row and then
+                        // allow the connector to contextualize it (i.e. perform any splitting)
+                        final CloseableIterator<FileReadContext> fileWithContext =
+                            parquetHandler.contextualizeFileReads(
+                                Utils.singletonCloseableIterator(
+                                    InternalUtils.getScanFileRow(nextFile)),
+                                Literal.TRUE);
+
+                        // Read that file
+                        final CloseableIterator<FileDataReadResult> fileReadDataIter =
+                            tableClient.getParquetHandler().readParquetFiles(
+                                fileWithContext,
+                                readSchema);
+
+                        return combine(fileReadDataIter, true /* isFromCheckpoint */);
+                    } else {
+                        throw new IllegalStateException(
+                            String.format("Unexpected log file path: %s", nextFilePath)
+                        );
+                    }
+                } catch (IOException ex) {
+                    throw new RuntimeException(ex);
```

Comment:
UncheckedIOException is an option, although I am not sure what the benefit of it.